apps/data-pipeline/course-scraper/src/index.ts

```diff
@@ -571,6 +572,317 @@ async function patchH32(meta: {
+
+  for (const yearData of sampleYears) {
+    const yearTerms: Array<{ term: string; courses: string[] }> = [
+      { term: "Fall", courses: [] },
+      { term: "Winter", courses: [] },
+      { term: "Spring", courses: [] },
+    ];
+
+    const curriculum = yearData.curriculum;
+
+    for (const row of curriculum) {
+      if (row.length >= 1 && row[0].trim()) {
+        yearTerms[0].courses.push(row[0].trim());
+      }
+
+      if (row.length >= 2 && row[1].trim()) {
+        yearTerms[1].courses.push(row[1].trim());
+      }
+
+      if (row.length >= 3 && row[2].trim()) {
+        yearTerms[2].courses.push(row[2].trim());
+      }
+    }
+
+    transformedProgram.push({ curriculum: yearTerms });
+  }
+
+  return {
+    sampleProgram: transformedProgram,
+  };
+}
+
+async function storeSampleProgramsInDB(
+  db: ReturnType<typeof database>,
+  scrapedPrograms: Array<{
+    programName: string;
+    sampleProgram: SampleProgramEntry[];
+    notes: string[];
+  }>,
+) {
+  if (!scrapedPrograms.length) {
+    logger.info("No sample programs to store.");
+    return;
+  }
+
+  logger.info("Fetching sample programs from database...");
+  const dbPrograms = await db
+    .select()
+    .from(sampleProgram)
+    .where(
+      inArray(
+        sampleProgram.programName,
+        scrapedPrograms.map((program) => program.programName),
+      ),
+    );
+
+  const sortedDbPrograms = sortKeys(dbPrograms, { deep: true });
+  const sortedScrapedPrograms = sortKeys(scrapedPrograms, { deep: true });
+
+  const programDiff = diffString(sortedDbPrograms, sortedScrapedPrograms);
+  if (!programDiff.length) {
+    logger.info("No difference found between database and scraped sample program data.");
+    return;
+  }
+
+  console.log("Difference between database and scraped sample program data:");
+  console.log(programDiff);
+  if (!readlineSync.keyInYNStrict("Is this ok")) {
+    logger.error("Cancelling sample program update.");
+    return;
+  }
+
+  await db.transaction(async (tx) => {
+    await tx.delete(sampleProgram).where(
+      inArray(
+        sampleProgram.programName,
+        scrapedPrograms.map((program) => program.programName),
+      ),
+    );
+
+    await tx.insert(sampleProgram).values(
+      scrapedPrograms.map((program) => ({
+        programName: program.programName,
+        sampleProgram: program.sampleProgram,
+        programNotes: JSON.stringify(program.notes),
+      })),
+    );
+  });
+
+  logger.info(`Successfully stored ${scrapedPrograms.length} sample programs`);
+}
+
+async function scrapeSamplePrograms(programPath: string) {
+  const url = `${CATALOGUE_URL}${programPath}`;
+  const html = await fetchWithDelay(url);
+  const $ = load(html);
+
+  const programName = $("h1.page-title").text().normalize("NFKD").split("(")[0].trim();
+  const sampleProgramContainer = $("#sampleprogramtextcontainer");
+
+  if (!sampleProgramContainer.length) return;
+
+  const sampleYears: { year: string; curriculum: string[][] }[] = [];
```

Comment:
you should decide whether to use `Array<T>` or `T[]` and stick that one throughout